Extra events for Doctrine QueryBuilder manipulation

tiraeth opened the pull request:
Extra events for easy (without the need to override findAll and findBy methods in AdminController) Doctrine QueryBuilder manipulation before passing it into Pagerfanta where it gets converted to Query and user is unable to modify it for his needs.

Use case? Let's say you want to split ``User`` entity into ``Administrators`` (users with ``ROLE_ADMIN`` role) and ``Users`` (users with ``ROLE_USER`` role). This will allow you to properly modify your ``QueryBuilder`` based on ``entity`` metadata. You are not required to override ``AdminController::findBy`` and ``AdminController::findAll`` methods anymore.

BCB? None.

ogizanagi commented:
I think it is a good idea, and the event is the simplest thing we can do. But actually I'm not sure this is the greatest thing to use an event for such kind of things.

I know the bundle's approach is far from tagged services & co, but I would love to see the `AdminController` less cluttered and have proper data providers and query extensions services for this (similar to what exists in the (DunglasApiBundle](https://github.com/dunglas/DunglasApiBundle/blob/master/Doctrine/Orm/DataProvider.php) if you already had the opportunity to have a look to it). 
It will perhaps even lead the way to an EasyAdminBundle able to support multiple data sources, not only Doctrine ORM. (But there is a lot of abstraction work that needs to be achieved for that)

I'm not against the event for now, but I think this is something we should think of in the future.

However, what about a simple new method in the `AdminController` class, that will allow to tweak the query builder instance ?

ogizanagi commented on Controller/AdminController.php:
```diff
@@ -528,6 +535,13 @@ protected function findBy($entityClass, $searchQuery, array $searchableFields, $
 
+        $this->dispatch(EasyAdminEvents::POST_SEARCH_QUERY_BUILDER, array(
+            'query_builder' => $queryBuilder,
+            'database_is_postgresql' => $databaseIsPostgreSql,
```
This is probably too specific for being part of an event IMO.

tiraeth replied:
Depends. If you insist on getting rid of this attribute — I can commit a fix and squash :-)

Pierstoval replied:
I also thing this parameter is useless. If your dbms is postgresql you should know it and therefore make queries that don't break postgresql features

tiraeth replied:
Unless you work on a publicly available bundle that will update query builder with some custom functionality. Anyway, I'll remove this option shortly :-)

tiraeth commented:
@ogizanagi I think the ``DataProvider`` approach will be nice to have in EasyAdminBundle. This depends on you guys, but I think that for now, this small event addon can help users be sure that when they update the library, things won't break. It gives the flexibility that is necessary here + don't introduce other philosophy and — IMO — is inline with how EasyAdminBundle currently works :-)

ogizanagi commented:
Right. But actually the bundle approach for now is mostly about extending the AdminController.

Let's ask @javiereguiluz and @Pierstoval about that :)

tiraeth commented:
@ogizanagi @javiereguiluz @Pierstoval added ``AdminController`` methods :-)

Pierstoval commented on Controller/AdminController.php:
```diff
@@ -487,6 +490,26 @@ protected function findAll($entityClass, $page = 1, $maxPerPage = 15, $sortField
+     * @param string      $sortDirection
+     * @param string|null $sortField
+     *
+     * @return Doctrine\ORM\QueryBuilder The Query Builder instance
```
Could you simplify the FQCN here?

Pierstoval commented on Controller/AdminController.php:
```diff
@@ -500,6 +523,36 @@ protected function findAll($entityClass, $page = 1, $maxPerPage = 15, $sortField
+     * @param string $searchQuery
+     * @param array  $searchableFields
+     *
+     * @return Doctrine\ORM\QueryBuilder The Query Builder instance
```
Same here for the FQCN than in my previous comment

Pierstoval commented:
I totally agree the methods you added, did not review in deep terms but the fact that the methods are "customizable" is great to me (I still want to "standardize" this to allow customizing any method for any entity)

tiraeth commented:
Rebased to single commit, fixed FQCN and missing ``return`` statement ;-) Also, I kept events for extendability in case someone wants to use them.

tiraeth commented:
@ogizanagi @Pierstoval I rebased with ``javiereguiluz/master`` for quicker merge. Also, updated ``AdminController`` to include #548 fix in both ``DoctrineORMAdapter `` constructors.

tiraeth commented:
And used ``executeDynamicMethod`` ;-)

javiereguiluz commented:
@tiraeth thanks for your first contribution and for adding this nice feature. Thanks also for your patience during the review process and for creating a perfectly rebased and ready-to-merge pull request.
